core/corehttp/gateway_handler.go

```diff
@@ -216,8 +227,32 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	parsedPath := ipath.New(urlPath)
-	if err := parsedPath.IsValid(); err != nil {
-		webError(w, "invalid ipfs path", err, http.StatusBadRequest)
+	if pathErr := parsedPath.IsValid(); pathErr != nil {
+		// Attempt to fix redundant /ipfs/ namespace as long as resulting
+		// 'intended' path is valid.  This is in case gremlins were tickled
+		// wrong way and user ended up at /ipfs/ipfs/{cid} or /ipfs/ipns/{id}
+		// like in bafybeien3m7mdn6imm425vc2s22erzyhbvk5n3ofzgikkhmdkh5cuqbpbq
+		// :^))
+		intendedPath := ipath.New(strings.TrimPrefix(urlPath, "/ipfs"))
+		if err := intendedPath.IsValid(); err == nil {
+			intendedURL := strings.Replace(r.URL.String(), urlPath, intendedPath.String(), 1)
+			// return HTML that
+			// - points at correct canonical path via <link> header
+			// - displays error
+			// - redirects to intendedURL after a delay
+			err = redirectTemplate.Execute(w, redirectTemplateData{
+				RedirectURL:   intendedURL,
+				SuggestedPath: intendedPath.String(),
+				ErrorMsg:      pathErr.Error(),
+			})
```

Comment:
This could be used for XSS. Our error messages usually don't allow arbitrary HTML. We likely need to do some escaping.

Reply:
we use `html/template` which takes care of escaping: 

```
<pre>invalid path: &#34;/ipfs/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR&#34; should be &#34;/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR&#34;</pre>
```

After https://github.com/ipfs/go-ipfs/pull/7930/commits/b81b7549d34e0b549c0035286bdd02b2be50030c  we also  pass string to template via  `%q`

Reply:
Got it. I assumed `text/template`.